src/tools/db-upgrade/DbUpgrader.cpp

```diff
@@ -0,0 +1,461 @@
+    auto sRet = schemaMan_->toGraphSpaceName(spaceId);
+    if (!sRet.ok()) {
+        LOG(INFO) << "space Id " << spaceId << " no found";
+        return sRet.status();
+    }
+    spaceName_ = sRet.value();
+    spaceId_   = spaceId;
+
+    auto spaceVidLen = metaClient_->getSpaceVidLen(spaceId_);
+    if (!spaceVidLen.ok()) {
+        return spaceVidLen.status();
+    }
+    spaceVidLen_ = spaceVidLen.value();
+
+    engine_.reset(new nebula::kvstore::RocksEngine(spaceId_, spaceVidLen_, FLAGS_upgrade_db_path));
+    auto parts = engine_->allParts();
+    parts_.clear();
+    parts_.insert(parts.begin(), parts.end());
+    tagSchemas_.clear();
+    tagIndexes_.clear();
+    edgeSchemas_.clear();
+    edgeIndexes_.clear();
+    return Status::OK();
+}
+
+
+Status DbUpgrader::buildSchemaAndIndex() {
+    auto tags = schemaMan_->getAllVerTagSchema(spaceId_);
+    if (!tags.ok()) {
+        LOG(INFO) << "space Id " << spaceId_ << " no found";
+        return tags.status();
+    }
+    tagSchemas_ = std::move(tags).value();
+
+    // Get all index in space
+    std::vector<std::shared_ptr<nebula::meta::cpp2::IndexItem>> tagIndexes;
+    auto iRet = indexMan_->getTagIndexes(spaceId_);
+    if (iRet.ok()) {
+        tagIndexes = std::move(iRet).value();
+    }
+
+    // handle index
+    for (auto&tagIndex : tagIndexes) {
+        if (tagIndex->schema_id.tag_id > 0) {
+            tagIndexes_[tagIndex->schema_id.tag_id].emplace(tagIndex);
+        }
+    }
+
+    // Get all edge in space
+    auto edges = schemaMan_->getAllVerEdgeSchema(spaceId_);
+    if (!edges.ok()) {
+        LOG(INFO) << "space Id " << spaceId_ << " no found";
+        return edges.status();
+    }
+    edgeSchemas_ = std::move(edges).value();
+
+    // Get all edge index in space
+    std::vector<std::shared_ptr<nebula::meta::cpp2::IndexItem>> edgeIndexes;
+    iRet = indexMan_->getEdgeIndexes(spaceId_);
+    if (iRet.ok()) {
+       edgeIndexes = std::move(iRet).value();
+    }
+    // handle index
+    for (auto& edgeIndex : edgeIndexes) {
+        if (edgeIndex->schema_id.edge_type > 0) {
+            edgeIndexes_[edgeIndex->schema_id.edge_type].emplace(edgeIndex);
+        }
+    }
+    return  Status::OK();
+}
+
+
+bool DbUpgrader::isValidVidLen(VertexID srcVId, VertexID dstVId) {
+    if (!NebulaKeyUtils::isValidVidLen(spaceVidLen_, srcVId, dstVId)) {
+        LOG(INFO) << "vertex id length is illegal, expect: " << spaceVidLen_
+                  << " result: " << srcVId << " " << dstVId;
+        return false;
+    }
+    return true;
+}
+
+// Used for vertex and edge
+std::string DbUpgrader::encodeRowVal(const RowReader* reader,
+                                     const meta::NebulaSchemaProvider* schema) {
+    auto oldSchema = reader->getSchema();
+    if (oldSchema == nullptr) {
+        LOG(INFO)  << "schema not found.";
+        return "";
+    }
+    std::vector<Value> props;
+    auto iter = oldSchema->begin();
+    size_t idx = 0;
+    while (iter) {
```

Comment:
`for (idx = 0; idx < oldSchema->getNumFields(); idx++)` ?

Reply:
Is equivalent